62_CAD/DrawResourcesFiller.cpp

```diff
@@ -2649,4 +2684,256 @@ void DrawResourcesFiller::flushDrawObjects()
+		pixelRatio = pixelRatio < 1.0 ? 1.0 : pixelRatio;
+
+		std::cout << "Sampled mip level: " << nbl::hlsl::log2(pixelRatio) << std::endl;
+	}
+		
+	// We need to make every tile that covers the viewport resident. We reserve the amount of tiles needed for upload.
+	core::vector<StreamedImageCopy> tiles;
+	uint32_t nTiles = 0;
+	for (uint32_t tileX = viewportTileRange.topLeft.x; tileX <= viewportTileRange.bottomRight.x; tileX++)
+		for (uint32_t tileY = viewportTileRange.topLeft.y; tileY <= viewportTileRange.bottomRight.y; tileY++)
+		{
+			// Compute tile offset relative to `currentMappedRegion.topLeft`, to get tile index into the gpu image
+			uint32_t2 gpuImageTileIndex = ((uint32_t2(tileX, tileY) - imageStreamingState->currentMappedRegion.topLeft) + imageStreamingState->gpuImageTopLeft) % imageStreamingState->gpuImageSideLengthTiles;
+			nTiles += !imageStreamingState->currentMappedRegionOccupancy[gpuImageTileIndex.x][gpuImageTileIndex.y];
+		}
+	tiles.reserve(nTiles);
+
+	// Assuming a 1 pixel per block format - otherwise math here gets a bit trickier
+	auto bytesPerPixel = getTexelOrBlockBytesize(imageStreamingState->georeferencedImageParams.format);
+	const size_t bytesPerSide = bytesPerPixel * GeoreferencedImageTileSize;
+
+	// Dangerous code - assumes image can be perfectly covered with tiles. Otherwise will need to handle edge cases
+	// TODO: All of this code only works for mip 0. Needs to be changed next to upload mip 1.
+	// Eventually this is all replaced by a few uploads to staging buffer + CS mip calc
+	for (uint32_t tileX = viewportTileRange.topLeft.x; tileX <= viewportTileRange.bottomRight.x; tileX++)
+	{
+		for (uint32_t tileY = viewportTileRange.topLeft.y; tileY <= viewportTileRange.bottomRight.y; tileY++)
+		{
+			// Compute tile offset relative to `currentMappedRegion.topLeft`, to get tile index into the gpu image
+			uint32_t2 gpuImageTileIndex = ((uint32_t2(tileX, tileY) - imageStreamingState->currentMappedRegion.topLeft) + imageStreamingState->gpuImageTopLeft) % imageStreamingState->gpuImageSideLengthTiles;
+
+			// If tile already resident, do nothing
+			if (imageStreamingState->currentMappedRegionOccupancy[gpuImageTileIndex.x][gpuImageTileIndex.y])
+				continue;
+
+			auto tile = georeferencedImageLoader->load(imageStreamingState->georeferencedImageParams.storagePath, uint32_t2(tileX, tileY) * GeoreferencedImageTileSize, uint32_t2(GeoreferencedImageTileSize, GeoreferencedImageTileSize), imageStreamingState->currentMappedRegion.baseMipLevel);
+
+			asset::IImage::SBufferCopy bufCopy;
+			bufCopy.bufferOffset = 0;
+			bufCopy.bufferRowLength = GeoreferencedImageTileSize;
+			bufCopy.bufferImageHeight = 0;
+			bufCopy.imageSubresource.aspectMask = IImage::EAF_COLOR_BIT;
+			bufCopy.imageSubresource.mipLevel = 0u;
+			bufCopy.imageSubresource.baseArrayLayer = 0u;
+			bufCopy.imageSubresource.layerCount = 1u;
+			uint32_t2 imageOffset = gpuImageTileIndex * GeoreferencedImageTileSize;
```

Comment:
name gpuImageOffset for clarity that this is not related to the large georef image but the resident one